Create operator bundle yaml for OSS and EE operators

mtyazici opened the pull request:
## Changes
- For hazelcast-operator, `hazelcastcluster.crd.yaml`, `operator-rbac.yaml` and `operator.yaml` are merged into one `bundle.yaml` file.

- For hazelcast-enterprise-operator two bundle files are created: `bundle.yaml` and `bundle-rhel.yaml`. The only difference between them is `bundle.yaml` contains `operator-docker-hub.yaml` while `bundle-rhel.yaml` contains `operator-rhel.yaml`.

- The release workflows are also affected by this change. 
  - For Docker Hub release, changes were minimal. Only file names are changed.

   - For Red Hat release, on top of file name changes test scripts had to be changed too. The command  `oc secrets link hazelcast-enterprise-operator pull-secret --for=pull` has to be run between operator service account creation and operator deployment. However with the `bundle.yaml`, there is no such possibility. So I chose to create a service account which is a clone of the one in `bundle-rhel.yaml` that is run before the aforementioned command. The order looks like the following


    ```
    
    cat <<EOF > ${WORKDIR}/service-account.yaml
    apiVersion: v1
    kind: ServiceAccount
    metadata:
      name: hazelcast-enterprise-operator
      labels:
        app.kubernetes.io/name: hazelcast-enterprise-operator
        app.kubernetes.io/instance: hazelcast-enterprise-operator
        app.kubernetes.io/managed-by: hazelcast-enterprise-operator
    EOF
    
    oc apply -f ${WORKDIR}/service-account.yaml
    
    oc apply -f ${WORKDIR}/hazelcast-rbac.yaml
    
    oc secrets link hazelcast-enterprise-operator pull-secret --for=pull
    oc apply -f ${WORKDIR}/bundle-rhel.yaml
    ```

     Bundle file also caused another problem when deleting the created resources. With the old script, Hazelcast CRD wasn't deleted after the test. Deleting `bundle-rhel.yaml` causes the CRD to be deleted but deletion does not complete successfully. So, in `clean-up.sh`, each resource created in `bundle-rhel.yaml` is deleted individually except the CRD.

leszko reviewed (commented):
@mtyazici I skimmed through this PR and I think there are still things we need to clarify (and the TDD is the right place for it), before reviewing it further.

In general, our development process should be:
1. You create TDD
2. TDD is reviewed, the solution is discussed, and the TDD is approved
3. Then we implement and send the PR

Otherwise, you may spend a lot of time on the implementation while we actually don't have an agreement on what we want to implement.

Some things that are not clear to me:
- Do we want to have ONLY `bundle.yaml`? Or do we want to have both, the files we had and the new `bundle.yaml`. The answer to this question is not that simple, because I see different operators do it in different ways. This requires research. If we decide to do both, then you don't need to change all these release files in this PR.
- I see sometimes bundle files are enriched with `openAPIV3Schema`. Do we want/need it?
- Sometimes bundle files are actually called `all-in-one.yaml` which is very explicit. Do we want to call it that way?
- Should Hazelcast RBAC be included in the bundle file? What do other operators do in similar situations?

I think we should well analyze (and later document it), because this will later be our recommended way of installing Hazelcast in k8/openshift.

mtyazici commented:
@leszko Thanks for the review. I will try to address your questions.


> In general, our development process should be:
>
>   1.You create TDD
>   2.TDD is reviewed, the solution is discussed, and the TDD is approved
>   3.Then we implement and send the PR
>
> Otherwise, you may spend a lot of time on the implementation while we actually don't have an agreement on what we want to implement.

I thought we had decided on the implementation when we said we would use one bundle file. From now on, I will make sure to start implementation after the TDD is approved.

> Do we want to have ONLY bundle.yaml? Or do we want to have both, the files we had and the new bundle.yaml. The answer to this question is not that simple, because I see different operators do it in different ways. This requires research. If we decide to do both, then you don't need to change all these release files in this PR.

I thought keeping the files that we merged into bundle would be redundant if we don't address them anywhere. When I updated the READMEs I also deleted those files as we were not using them anywhere. However, I can look at what other projects are doing, then we can decide on what to do.

> I see sometimes bundle files are enriched with openAPIV3Schema. Do we want/need it?

openAPIV3Schema is used for validating the created resource at the Kubernetes API. I think it is valuable to have the schema defined. But of course we can discuss its need.

> Sometimes bundle files are actually called all-in-one.yaml which is very explicit. Do we want to call it that way?

As far as I saw, only a few operators' bundle file are called `all-in-one.yaml`. I saw others using `bundle.yaml` and `operator.yaml`. I don't think we should change the bundle file's name.

> Should Hazelcast RBAC be included in the bundle file? What do other operators do in similar situations?

In my research I haven't come across with bundle files that include RBAC for operator resources. However, those bundle files contained a lot CRDs. Currently, our operator only monitors one custom resource, so I think it is reasonable to also add the RBAC for Hazelcast resource type. 

I found the [Spark project](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator) that creates the operator resource RBAC as the operator is deployed, though they are using [Helm Chart](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/tree/master/charts/spark-operator-chart) instead of bundle file. The similarity between our and their operator is that both operators only watch one custom resource. So, it would also make sense for us to add RBAC for Hazelcast resource in the bundle file.

leszko commented on .github/scripts/clean-up.sh:
```diff
@@ -7,10 +7,16 @@ set -o pipefail
+oc delete -f  ${WORKDIR}/secret.yaml --wait=true
+oc delete -f  ${WORKDIR}/hazelcast-rbac.yaml --wait=true
+
+oc delete --wait=true rolebinding hazelcast-enterprise-operator
```
Won't it just work with `oc delete -f bundle.yaml`?

mtyazici replied:
If there are CRs existing in other namespaces when we try to delete the CRD, the operation fails and script gets stuck deleting the CRD. So to solve this, we always apply the CRD in `bundle.yaml` whether it is already there or not, however we never delete the CRD to circumvent the situation I mentioned above. Those seperate `delete` commands are the result of this problem.

There is an issue about it https://github.com/elastic/cloud-on-k8s/issues/1701, at the bottom there is a link to another issue that show the temporary solution to the problem.

leszko commented on .github/scripts/smoke-test.sh:
```diff
@@ -29,12 +29,23 @@ oc create secret docker-registry pull-secret \
  --docker-email=$RED_HAT_EMAIL
 
-oc apply -f ${WORKDIR}/operator-rbac.yaml
+cat <<EOF > ${WORKDIR}/service-account.yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: hazelcast-enterprise-operator
+  labels:
+    app.kubernetes.io/name: hazelcast-enterprise-operator
+    app.kubernetes.io/instance: hazelcast-enterprise-operator
+    app.kubernetes.io/managed-by: hazelcast-enterprise-operator
+EOF
+
+oc apply -f ${WORKDIR}/service-account.yaml
```
Nit

```suggestion
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ServiceAccount
metadata:
  name: hazelcast-enterprise-operator
  labels:
    app.kubernetes.io/name: hazelcast-enterprise-operator
    app.kubernetes.io/instance: hazelcast-enterprise-operator
    app.kubernetes.io/managed-by: hazelcast-enterprise-operator
EOF
```

mtyazici replied:
Looks a lot better, I will change it. Thanks :)

leszko reviewed (commented):
Added 2 comments, other than that it looks good to me.

But wait also for @hasancelik to review.

leszko commented on hazelcast-enterprise-operator/README.md:
````diff
@@ -191,9 +172,24 @@ Insert this value into `secret.yaml`, replace `<base64-hz-license-key>`. Then, c
+Before starting the cluster, you need to remove the `securityContext` field from `hazelcast.yaml`.
+
+
+```
+...
+      memory: 1024Mi
+ securityContext:
+    runAsUser: ""
+    runAsGroup: ""
+    fsGroup: ""
+  mancenter:
+    image:
+...
+```
````
```suggestion
```
 securityContext:
    runAsUser: ""
    runAsGroup: ""
    fsGroup: ""
```
```

leszko commented on hazelcast-enterprise-operator/README.md:
```diff
@@ -191,9 +172,24 @@ Insert this value into `secret.yaml`, replace `<base64-hz-license-key>`. Then, c
+#### Step 4: Start Hazelcast
 
-Start Hazelcast cluster with the following command.
+Before starting the cluster, you need to remove the `securityContext` field from `hazelcast.yaml`.
```
```suggestion
Before starting the cluster, you need to remove the `securityContext` part from `hazelcast.yaml`.
```

leszko commented on hazelcast-operator/README.md:
```diff
@@ -93,31 +84,37 @@ Note: You need to clone this repository before following the next steps.
+#### Step 3: Start Hazelcast
 
-    kubectl --validate=false apply -f operator.yaml
+Before starting the cluster, you need to remove the `securityContext` field from `hazelcast.yaml`.
```
```suggestion
Before starting the cluster, you need to remove the `securityContext` part from `hazelcast.yaml`.
```

leszko commented on hazelcast-operator/README.md:
````diff
@@ -93,31 +84,37 @@ Note: You need to clone this repository before following the next steps.
 
-#### Step 4: Start Hazelcast
+```
+...
+      memory: 1024Mi
+ securityContext:
+    runAsUser: ""
+    runAsGroup: ""
+    fsGroup: ""
+  mancenter:
+    image:
+...
````
```suggestion
 securityContext:
    runAsUser: ""
    runAsGroup: ""
    fsGroup: ""
```

hasancelik reviewed (approved):
👍
